File: rustdsl/classes/tests/mixin/mixin.dart
abstract class I {
  void i();
}

abstract class A {
  final int x = 0;
  void f() {
    print('A::f, x = $x');
  }
}

abstract class B extends A {
  final int y = 1;
  @override
  void f() {
    super.f();
    print('B::f, y = $y');
  }

  void g() {
    print('B::g, y = $y');
  }
}

mixin M on A implements I {
  final int z = 2;
  @override
  void f() {
    super.f();
    print('M::f, z = $z');
  }

  void h() {
    print('M::h, z = $z');
  }
}

class C1 extends A with M implements I {
  final w = 3;
  @override
  void f() {
    super.f();
    print('C1::f, w = $w');
  }

  @override
  void h() {
    super.h();
    print('C1::h, w = $w');
  }

  @override
  void i() {
    print('C1::i, w = $w');
  }

  void j() {
    print('C1::j, w = $w');
  }
}

class C2 extends B with M implements I {
  final v = 4;
  @override
  void f() {
    super.f();
    print('C2::f, v = $v');
  }

  @override
  void g() {
    super.g();
    print('C2::g, v = $v');
  }

  @override
  void h() {
    super.h();
    print('C2::h, v = $v');
  }

  @override
  void i() {
    print('C2::i, v = $v');
  }

  void j() {
    print('C2::j, v = $v');
  }
}

void main() {
  final c1 = C1();
  c1.f();
  c1.h();
  c1.i();
  c1.j();

  M c1m = c1;
  c1m.f();
  c1m.h();
  print("z = ${c1m.z}");

  final c2 = C2();
  c2.f();
  c2.g();
  c2.h();
  c2.i();
  c2.j();

  M c2m = c2;
  c2m.f();
  c2m.h();
  print("z = ${c2m.z}");
}

File: rustdsl/classes_macros/tests/vtable_layout/mixin.rs
use pretty_assertions::assert_eq;

::classes::classes! {
    class A {
        pub fn new() -> Self { Self {} }
        pub fn f(&self) {}
    }
    abstract class I {
        pub fn i(&self);
    }
    #[with(A)]
    mixin M1 implements I {
        pub fn g(&self) {}
    }
    #[with(A/M1)]
    mixin M2 on M1 {
        pub override fn M1::g(&self) {}
        pub          fn     h(&self) {}
        pub override fn <M1 as I>::i(&self) {}
    }
    class B extends A with M1, M2 {
        pub fn new() -> Self { Self { super: Super::new(), .. } }
        pub override fn  A::f(&self) {}
        pub override fn M1::g(&self) {}
        pub override fn M2::h(&self) {}
    }
}

#[test]
fn test_mixin() {
    assert_eq!(
        format!("{:#?}", B::vtable().debug_vtable_layout()),
        "(
    [
        MixinVtableHeader {
            instance: vtable_layout::mixin::A_M1_M2,
            mixin_offset: 48,
            super_offset: 56,
            data_offset: 0,
        },
        MixinVtableHeader {
            instance: vtable_layout::mixin::A_M1,
            mixin_offset: 24,
            super_offset: 24,
            data_offset: 0,
        },
    ],
    B {
        'start: 0,
        super: A_M1_M2 {
            'start: 0,
            super: A_M1 {
                'start: 0,
                super: A {
                    'start: 0,
                    super: Object {
                        'start: 0,
                        header: ObjectHeader {
                            object_ty: vtable_layout::mixin::B,
                        },
                        'end: 16,
                    },
                    f: 16,
                    'end: 24,
                },
                g: 24,
                I: I {
                    'start: 32,
                    super: Object {
                        'start: 32,
                        header: ClassHeader {
                            class_ty: vtable_layout::mixin::I,
                            offset: -32,
                        },
                        'end: 48,
                    },
                    i: 48,
                    'end: 56,
                },
                'end: 56,
            },
            h: 56,
            'end: 64,
        },
        'end: 64,
    },
)",
    );

    let b = B::new();
    assert_eq!(
        format!(
            "{:#?}",
            ::classes::class::ClassRcWeak::vtable(&b).debug_vtable_layout(0)
        ),
        "B {
    'start: 0,
    super: A_M1_M2 {
        'start: 0,
        super: A_M1 {
            'start: 0,
            super: A {
                'start: 0,
                super: Object {
                    'start: 0,
                    header: ObjectHeader {
                        object_ty: vtable_layout::mixin::B,
                    },
                    'end: 16,
                },
                f: 16,
                'end: 24,
            },
            g: 24,
            I: I {
                'start: 32,
                super: Object {
                    'start: 32,
                    header: ClassHeader {
                        class_ty: vtable_layout::mixin::I,
                        offset: -32,
                    },
                    'end: 48,
                },
                i: 48,
                'end: 56,
            },
            'end: 56,
        },
        h: 56,
        'end: 64,
    },
    'end: 64,
}",
    );

    let m2: CRc<M2> = b.into_supertype();
    assert_eq!(
        format!(
            "{:#?}",
            ::classes::class::ClassRcWeak::vtable(&m2).debug_vtable_layout()
        ),
        "(
    [
        MixinVtableHeader {
            instance: vtable_layout::mixin::A_M1_M2,
            mixin_offset: 48,
            super_offset: 56,
            data_offset: 0,
        },
        ..
    ],
    M2 {
        'start: 56,
        h: 56,
        'end: 64,
    },
)"
    );

    let m1: CRc<M1> = m2.into_supertype();
    assert_eq!(
        format!(
            "{:#?}",
            ::classes::class::ClassRcWeak::vtable(&m1).debug_vtable_layout()
        ),
        "(
    [
        MixinVtableHeader {
            instance: vtable_layout::mixin::A_M1,
            mixin_offset: 24,
            super_offset: 24,
            data_offset: 0,
        },
        ..
    ],
    M1 {
        'start: 24,
        g: 24,
        I: I {
            'start: 32,
            super: Object {
                'start: 32,
                header: ClassHeader {
                    class_ty: vtable_layout::mixin::I,
                    offset: -32,
                },
                'end: 48,
            },
            i: 48,
            'end: 56,
        },
        'end: 56,
    },
)"
    );
}
